src/indieweb_utils/feeds/discovery.py

```diff
@@ -67,3 +68,41 @@ def discover_web_page_feeds(url: str, user_mime_types: Optional[List[str]] = Non
         feeds.append(FeedUrl(url=url, mime_type="text/html", title=page_title.text))
 
     return feeds
+
+
+def discover_h_feed(url: str) -> dict:
+    """
+    Find the main h-feed that represents a web page as per the h-feed Discovery algorithm.
+
+    refs: https://microformats.org/wiki/h-feed#Discovery
+
+    :param url: The URL of the page whose associated feeds you want to retrieve.
+    :type url: str
+    :return: The h-feed data.
+    :rtype: dict
+    """
+
+    parsed_main_page_mf2 = mf2py.parse(url=url)
+
+    all_page_feeds = discover_web_page_feeds(url)
+
+    if all_page_feeds.get("text/mf2+html"):
+        # url has already been canonicalized by discover_web_page_feeds
+        # thus, no canonicalization is needed in this function
+
+        feed = all_page_feeds.get("text/mf2+html").url
+
+        parsed_feed = mf2py.parse(url=feed)
+
+        h_feed = [item for item in parsed_feed["items"] if item.get("type", [])[0] == "h-feed"]
```

Comment:
If there is no `"type"` and the default parameter `[]` is returned, this code will throw an `IndexError` because you can't access item 0 of an empty array. Instead we should check that item.get("type") exists (and None will evaluate as False)  and then if the type of the first element equals "h-feed".

```suggestion
        h_feed = [item for item in parsed_feed["items"] if item.get("type") and item.get("type")[0] == "h-feed"]
```

Reply:
@capjamesg We should apply this suggestion (or something similar) before merge to prevent a potential crasher.